src/scraper/tests/el_pitazo_test.py

```diff
@@ -0,0 +1,23 @@
+# TODO move tests out this module when approved.
+
+from scraper.tests.utils import fake_response_from_file
+from scraper.scrapers.el_pitazo_scraper import ElPitazoScraper
+from scraper.settings import ROOT_DIR
+import os 
+
+
+def test_title_parse_ok():
+    url = "tests/html_bodies/el_pitazo_fallas_electricas_carne.html"
+    test_file = os.path.join(ROOT_DIR, url)
+    response = fake_response_from_file(test_file, "https://elpitazo.net/cronicas/la-fallas-electricas-han-disminuido-la-cantidad-de-carne-que-consume-el-venezolano/")
```

Comment:
nice 👍🏽

Reply:
As I said to Luis, I would not even test that the website is correctly parsed. Why? Website layouts are fragile, and fragile tests are bad tests.

There needs to be a separation of concern: one thing is parsing correctly, and another one is using the parser correctly. We need to test the latter because that's the purpose of a spider—among other things. Creating these kinds of tests for one spider is simple, but as websites are added and they change, it will be a hell to maintain.

Instead, we should (eventually) prepare the system for monitoring and alerts, but I would not worry about that—for now.

Reply:
> Why? Website layouts are fragile, and fragile tests are bad tests.

In my opinion, if the code is expected to go read a webpage and extract information from it - there should be at least a test asserting that if the website looks one way, then X information is extracted from it. Not sure how else you would test that your scraper is working correctly on one version of the intended website. When the website changes layout, then we can change the test html we've saved for that version. 

If way more websites are added and this becomes complex, then we can refactor our approach. 

What other approaches do you suggest for testing this?

Reply:
> and another one is using the parser correctly. We need to test the latter because that's the purpose of a spider—among other things.
not sure I get what you mean by `using the parser correctly` :)

Reply:
Avoiding testing fragile code. 

Just like you would (usually) not test a UI. Integration and E2E tests are expensive and a hell to maintain because they are fragile. I would apply [the Humble Object pattern](https://martinfowler.com/bliki/HumbleObject.html). Make the parser humble: its only job is to execute a parsing function (that has already been tested, e.g. beautifulsoup, lxml, ...). We make sure that the parser works (pass simple HTML strings with different selectors) and then move on.

The interesting logic we want to check is not whether something has been correctly configured or not. We want to assert that our program can configure itself based on the inputs (now _that_ is easy to test). For example, URL starts with pitazo? Pitazo scraper it is. The Pitazo scraper does not return anything because the selectors are wrong? Well, that's the problem of the configuration, not the class. The class works. Its behavior is correct. No need for file fixtures. No need to update both code and tests every time a single `div` changes or when a CSS/XPath selector is not general enough.

When the website changes layout (how are you going to know that it changed?), our tests must not break. An alert system should. We should not couple a whole Python library to a website CSS and HTML.

Reply:
> In my opinion, if the code is expected to go read a webpage and extract information from it [...]

The code is expected to configure a spider appropriately based on the URL. It might be expected to clean the output a bit assuming a correct parsing. It is expected to do something when no input was found (because of connection errors or simple bad CSS selectors or layout changes), for example, just write an empty string to a Dict.

But, I do not expect the code to extract information from a webpage with 100% accuracy just like I do not expect an NLP algorithm to classify text perfectly. That's why we should validate our scraping engine with a monitoring system just like a Machine Learning Pipeline is constantly being evaluated and re-trained. We can only test deterministic algorithms. Scraping something that is not under our control is not one of those.

Reply:
> When the website changes its layout (how are you going to know that it changed?), our tests must not break.

Our tests won't break, since that one sample page is stored locally in the project. 
In the case of a **mostly static website**, where most pages have the same layout (chatting with Luis, looks like el pitazo is mostly on this line) - I think the value of the integration test will outweigh the maintenance burden of it (which should be very little).

If that website changes (and all pages adopt a new layout) - then we'll have to update our parser anyways. (I'm with you on the fact that an alerting system should warn us of this, not our tests). Once we're aware the site changed, then we'd take a snapshot of what the new layout (hopefully still as static) is like for an article page, and update our test with it. 

**Of course,** in the case of a more complicated website, or one where the set up is more complex - the overhead of maintenance increases a lot and the overall value therefore decreases. In which case I agree with you more we shouldn't aim to write such tests. 

Now there may be some scrappy specifics that make this impractical - about these I'm all ears :) 

My end impression is simply that for a website whose pages are static and follow the same layout - there is more value than cost in this kind of snapshot based integration test.